### DISCUSSION TITLE
how to implement classical fourth-order Runge–Kutta scheme with TCLB?

### DISCUSSION THREAD

**Discussion by thebetter2** [CFD-GO/TCLB#469] [General]

**Comment by llaniewski** [comment]
@thebetter2 did you manage to solve the problem? I think the best person to talk to wrt. Runge-Kutta in TCLB is @TravisMitchell , as he made the initial implementation.

**Reply by thebetter2** [reply]
Thank you. The reason why I closed the issue is that I later solved RK-4 (sorry, I closed discuss for the first time, so the content was deleted due to improper operation). I read the following website https://develop.docs.tclb.io/tutorials/model-development/1.-finite-difference-wave-equation/ repeatedly. I found that my previous understanding of stage was wrong. The solution for RK4 can be achieved by changing it to the following:
```
AddField("psi", stencil3d = 2)
AddField("temper", stencil3d = 1,group="tt")
AddField("t_ux", stencil3d = 1,group="t")
AddField("t_uy", stencil3d = 1,group="t")
AddField("t_uz", stencil3d = 1,group="t")
AddField("t_lambda", stencil3d = 1,group="t")
AddField("rk1", stencil3d = 1,group="Trk1")
AddField("rk2", stencil3d = 1,group="Trk2")
AddField("rk3", stencil3d = 1,group="Trk3")
AddField("rk4", stencil3d = 1,group="Trk4")

# Stages and Actions
AddStage("R_K1","calcT_h1" , save="rk1", load=DensityAll$group %in% c("f","density","t","tt"))
AddStage("R_K2","calcT_h2" , save="rk2", load=DensityAll$group %in% c("f","density","t","tt","Trk1"))
AddStage("R_K3","calcT_h3" , save="rk3", load=DensityAll$group %in% c("f","density","t","tt","Trk1","Trk2"))
AddStage("R_K4","calcT_h4" , save="rk4", load=DensityAll$group %in% c("f","density","t","tt","Trk1","Trk2","Trk3"))
AddStage("PsiIteration", "calcPsi"    , save="psi", load=DensityAll$group %in% c("f","density","tt") )
AddStage("BaseInit"     , "Init", save=Fields$group %in% c("f","density","t","tt"))
AddStage("BaseIteration", "Run",  save=Fields$group %in% c("f","density","t","tt"), load=DensityAll$group %in% c("f","density","t","tt"))

AddAction("Init"     , c("BaseInit" ,"PsiIteration","R_K1","R_K2","R_K3","R_K4")) 
AddAction("Iteration", c("BaseIteration", "PsiIteration","R_K1","R_K2","R_K3","R_K4"))
```
I have just used TCLB for a while. It is very good, has a high degree of customization and good flexibility.

**Reply by llaniewski** [reply]
@thebetter2 Good to hear.

We're always interested in hearing who uses TCLB and for what, as people using it makes the project worthwhile. If you can, please drop me an email about what you're doing (lukasz.laniewski@pw.edu.pl or l.laniewskiwollk@uq.edu.au). If this is a part of a publication or a thesis, maybe you would want it added to the https://docs.tclb.io/general-info/publications/ ?

**Comment by TravisMitchell** [comment]
@thebetter2 I did the RK4 implementation to couple thermal and multiphase effects in the phase field LBM (d3q27_pf_velocity), so would be happy to help if you need at all :).

**Reply by thebetter2** [reply]
Thank you. I just solved RK4 by referring to the d3q27_pf_velocity model, combined with my repeated reading of the tutorial.
